### ISSUE TITLE
Hairpin NAT

### ISSUE THREAD

**Issue by vincent99** [rancher/rancher#1920]
Requests from the host to the public IP and back to the host do not work.

**Comment by deniseschannon** [comment]
There have been many users hitting this issue (either on forums or in GitHub).

**Comment by kaos** [comment]
+1

**Comment by andreycizov** [comment]
Any case there's a temporary fix for this problem as a set of iptables rules?

**Comment by flaccid** [comment]
+1
Would like a workaround for now if possible please.

**Comment by y8** [comment]
@andreycizov check #2929 for workaround.

**Comment by y8** [comment]
Just in case, I'll repost from #2929

TL:DR: See: https://github.com/y8/rancher-localroute. This is temporary solution until issue is resolved by Rancher. Enable `net.ipv4.conf.all.route_localnet` and `net.ipv4.ip_forward`, reboot host and then fire this compose file on all your environments:

````yml
rancher-localroute:
  labels:
    io.rancher.scheduler.global: 'true'
  tty: true
  image: yopp/rancher-localroute:latest
  privileged: true
  stdin_open: true
  net: host
````

Long part:

This issue is huge pain in the ass: _all_ the services on the same host can't talk to each other using the public IP. Route53 for service discovery? Ha-ha. Good luck debugging randomly failed services with cryptic "connection refused" errors. I don't know why #147 was closed, but this is must be addressed as soon as possible.

There two options here:

1. Opt-out style `docker-proxy` option for exposed ports.
2. Mandatory `route_localnet` on host system and two extra iptables rules per port: one to route locally generated packages to the IPSec network (OUTPUT DNAT), and one to change source address from loopback to //some// IPSec routable address, so target container can respond to the real source host, not the loopback (PREROUTING SNAT).

  Example:

  ````
-t nat -A POLLER_OUTPUT
       -d %{host_public_ip}
       -p %{proto} 
       --dport %{source_port} 
       -j DNAT 
       --to %{dest_address}:%{dest_port}"

-t nat -A POLLER_POSTROUTING
       -d %{dest_address} 
       -p %{proto} 
       --dport %{dest_port} 
       -j SNAT 
       --to-source %{host_public_ip}"
  ````

I don't know anything about possible side-effecs of enabling `route_localnet`, but it seems like the best solution, because it requires couple extra lines in the https://github.com/rancher/cattle/blob/master/resources/content/config-content/host-iptables/content-home/etc/cattle/host-iptables.ftl. 

Additionally `route_localnet` flag must be verified upon network-agent start. On our ubutnu hosts changing the `net.ipv4.conf.all.route_localnet` had no effect until the reboot. I've tried restarting networking service, but without any luck. It might be related to the `veth` interface management.

I'm bad on shell scripting, so I've made a small container with simple ruby script inside. It polls the `CATTLE_PREROUTING` chain and updates `POLLER_OUTPUT` and `POLLER_PREROUTING` accordingly. See https://github.com/y8/rancher-localroute. Quick and dirty, but works on our ubuntu cluster.

Cheers.

**Comment by y8** [comment]
I've released the `yopp/rancher-localroute:0.2.0` with multiple host ip support and some additional configuration options. 

Check the README for details: https://github.comy8/rancher-localroute.

**Comment by ibuildthecloud** [comment]
Okay, I believe in 0.52 this will be addressed.  What we are going to do is just pass the port info on to Docker.  This issue has caused enough problems for users that that is the best short term solution we believe.

The problem with Docker managing the ports is that we can't dynamically change them.  Also to get localhost to work Docker uses a user space proxy which is slow.  There is the non user space proxy which requires route_localnet but that seems to be causing kernel issues with users.

**Comment by Mike-Petersen** [comment]
+1
Hope 0.52 will be release soon. Unfortunately the workaround does not seem to work anymore.

**Comment by djskinner** [comment]
Might this be the reason I cannot deploy containers to a host if that host is also running the my private registry?

**Comment by djskinner** [comment]
Is this resolved with the new networking in  0.56?

**Comment by deniseschannon** [comment]
@djskinner No, this is not yet resolved as I've tried testing it and still not working.

**Comment by vincent99** [comment]
One more release.. https://github.com/rancher/cattle/pull/1302

**Comment by deniseschannon** [comment]
Verified in v0.57.0-rc1 in Ubuntu, RancherOS, RHEL 7.2

Verification Steps
1. Launch an nginx service with expose port on host
2. Exec into nginx container
3. `curl http://<ip_of_host>:<exposed_port>`

**Comment by ndarilek** [comment]
Is there some sort of upgrade step I need to perform above and beyond upgrading to Rancher 0.59? This isn't working for me, or at least, not in the way I'd expect.

http://vm1.mydomain is fronted by a Rancher load balancer. If I hit the URL from anywhere other than when SSH'd into vm1.mydomain, it works just fine. It doesn't work when actually SSH'd into vm1.mydomain, though.

Do I need to delete and recreate my load balancer?

**Comment by kaos** [comment]
Make sure your Load balancer has been upgraded to the newer image.

For me, it didn't upgrade the load balancer automatically, but when I deleted the network agent container, it got replaced, and with it the load balancer was also upgraded.

I can also confirm that this upgrade does solve the hairpin issue without any extra action besides making sure it gets the new image, so if you're still having issues, I'd think it's likely caused by something else..

**Comment by ndarilek** [comment]
I'm sorry to say this is not working for me.

I've removed all containers on my host, deleted the host from Rancher 
and re-added it. I cannot hit the load balancer IP directly when SSH'd 
into the host.

The agents appear to be upgraded, running 0.8.0 and 0.9.2. Somehow 
before I was still stuck on 0.4.1, but now everything appears updated.

How do I debug further?

**Comment by kaos** [comment]
Hmm.. not sure what's going on, but I seem to be getting hairpin issues still. We're investigating, but somehow, what seemed to work right after the upgrade doesn't work any more. So either I'm being confused now, or I was confused before...

**Comment by kaos** [comment]
I think the hairpin issue might actually have been fixed, but other stuff seems rather broken:
```
CONTAINER ID        IMAGE                                                                 COMMAND                  CREATED             STATUS              PORTS                NAMES
dc95ca48dd18        rancher/agent-instance   "/etc/init.d/agent-in"   2 hours ago         Up 2 hours          0.0.0.0:80->80/tcp   4f2d427a-2d1b-46c2-b1cf-260365c29977
[root@xxx ~]# curl -v 127.0.0.1:80
* About to connect() to 127.0.0.1 port 80 (#0)
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 127.0.0.1
> Accept: */*
> 
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

I am however using docker 1.8.2, so will upgrade to 1.10 first and see if it works better then..

Here's my NAT rules on the host in question, if it is of any help.. @ibuildthecloud do you spot any thing off here?
```
[root@tsy01 ~]# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
CATTLE_PREROUTING  all  --  anywhere             anywhere            
DOCKER     all  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere            !loopback/8           ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
CATTLE_POSTROUTING  all  --  anywhere             anywhere            
MASQUERADE  all  --  172.17.0.0/16        anywhere            
MASQUERADE  udp  --  172.17.0.1           172.17.0.1           udp dpt:ipsec-nat-t
MASQUERADE  udp  --  172.17.0.1           172.17.0.1           udp dpt:isakmp
MASQUERADE  tcp  --  172.17.0.4           172.17.0.4           tcp dpt:http

Chain CATTLE_POSTROUTING (1 references)
target     prot opt source               destination         
ACCEPT     all  --  10.42.0.0/16         169.254.169.250     
MASQUERADE  tcp  --  10.42.0.0/16        !10.42.0.0/16         masq ports: 1024-65535
MASQUERADE  udp  --  10.42.0.0/16        !10.42.0.0/16         masq ports: 1024-65535
MASQUERADE  all  --  10.42.0.0/16        !10.42.0.0/16        
MASQUERADE  tcp  --  172.17.0.0/16        anywhere             masq ports: 1024-65535
MASQUERADE  udp  --  172.17.0.0/16        anywhere             masq ports: 1024-65535

Chain CATTLE_PREROUTING (1 references)
target     prot opt source               destination         
DNAT       tcp  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL tcp dpt:http to:10.42.35.12:80

Chain DOCKER (2 references)
target     prot opt source               destination         
DNAT       udp  --  anywhere             anywhere             udp dpt:ipsec-nat-t to:172.17.0.1:4500
DNAT       udp  --  anywhere             anywhere             udp dpt:isakmp to:172.17.0.1:500
DNAT       tcp  --  anywhere             anywhere             tcp dpt:http to:172.17.0.4:80
```

**Comment by deniseschannon** [comment]
I just tested and created a separate issue. https://github.com/rancher/rancher/issues/3618

**Comment by qrpike** [comment]
+1

**Comment by djskinner** [comment]
I've upgrades to 0.59.1 and am still having issues. Are there still known issues related to this?

Basically I have a load balancer listening on port 80. Traffic for mydomain.com gets routed to a container on host 1. If I SSH into host 1 and `curl mydomain.com` I get `curl: (7) Failed to connect to mydomain port 80: Connection refused`

If I do the same curl from host 2 I don't get the connection refused error.

**Comment by kaos** [comment]
@djskinner that is #3618
